### PR TITLE
add inMemory Backend

### LINKED ISSUE #152: In-memory mosaic backend
If you have a mosaic in memory and you want to access backend methods like `.tile`, you currently need to know that you can pass `FileBackend(path=None, mosaic_def=...)`. It might be clearer to expose a `InMemoryBackend` that doesn't require a `path` argument.

### PR DESCRIPTION
**Pull request by vincentsarago**
closes #152 

I'm not sure which `flow` to choose

```python
# This PR
with InMemoryBackend(mosaicjson) as mosaic:
    ...
```

or 

```python
with InMemoryBackend(":memory:") as mosaic:
    mosaic.write(mosaicjson)
    ...
```

The second block will feel more aligned with how the other backend works, but also feels kinda weird.

Note: I didn't add the InMemoryBackend in the Top level MosaicBackend because I wasn't 💯 sure we wanted this.

### CODE REVIEW

**Comment by kylebarron** [comment]
Do we still allow passing `mosaic_def` in the class constructor to other backends? If so, it makes sense to do it here, but maybe require it to be a keyword argument?

**Comment by vincentsarago** [comment]
> Do we still allow passing mosaic_def in the class constructor to other backends? If so, it makes sense to do it here, but maybe require it to be a keyword argument?

@kylebarron no we don't, for the other backend, mosaic_def is outside the `init` and must be set using `.write(mosaic)`. IMO the InMemoryBackend is a bit different than the other backend which is why I'm fine to not follow the same pattern as the other (meaning, not passing a path and not using the .write method).

🤷‍♂️

File: cogeo_mosaic/backends/memory.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import attr
+from morecantile import TileMatrixSet
+from rio_tiler.constants import WEB_MERCATOR_TMS
+from rio_tiler.io import BaseReader, COGReader
+
+from cogeo_mosaic.backends.base import (
+    BaseBackend,
+    mode_validator,
+    update_zoom_and_bounds,
```
Comment (kylebarron) [inline cogeo_mosaic/backends/memory.py:13]: See this seems weird. Why not just put `mode_validator` and `update_zoom_and_bounds` on the class?

File: cogeo_mosaic/backends/memory.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        self.bounds = self.mosaic_def.bounds
+
+    def write(self, mosaic: MosaicJSON, overwrite: bool = True):
+        """Write mosaicjson document."""
+        if not isinstance(mosaic, MosaicJSON):
+            mosaic = MosaicJSON(**dict(mosaic))
+        self.mosaic_def = mosaic
+
+    def _read(self) -> MosaicJSON:
+        raise NotImplementedError("InMemoryBackend is a Write-Only backend.")
```
Comment (kylebarron) [inline cogeo_mosaic/backends/memory.py:62]: 😄

**Review by kylebarron — APPROVED** [review]

**Pull request closed without merging**